### PR TITLE
[DO NOT MERGE] Reference refactoring of BaseInnerProductAttributor

### PR DESCRIPTION
**Pull request by jiaqima**
## Description

This PR is a reference implementation to refactor the BaseInnerProductAttributor, which is NOT intended to be merged (and has not been tested). As an example, the Arnoldi attributor is re-implemented to demonstrate how the new API of BaseInnerProductAttributor can be used.

A few major changes in BaseInnerProductAttributor:
- transformation could now be applied to both train data representations/gradients and test data representations/gradients
- the batching over the full training data for transformation was original dealt inside `.attribute()` with a simple average, which lacks enough flexibility; it is now moved into the transform functions
- `.cache()` could do more than just storing the full training dataloader. For example, the Arnoldi projector is now processed and cached in transformation_on_query in a hacky way. In this PR, this is moved to .cache()
- removed transformation_kwargs and added an __init__ for the subclass. It's better to specify these attributor hyperparameters explicitly than relying on a kwargs

### CODE REVIEW

File: dattri/algorithm/base.py [hunk 1]
```diff
@@ -117,104 +108,127 @@ def _set_full_train_data(self, dataloader: torch.utils.data.DataLoader) -> None:
...
     ) -> torch.Tensor:
-        """Calculating the query based on the train data.
+        """Generate the initial representations of the train data.
 
-        Inner product attributor calculates the inner product between the
-        train query and the transformation of the test query. This function
-        calculates the train query based on the train data.
+        Inner product attributor calculates the inner product between the (transformed)
+        train representations and test representations. This function generates the
+        initial train representations.
```
Comment (TheaperDeng) [inline dattri/algorithm/base.py:150]: The default implementation calculates the gradient of the train loss with respect to the parameter. Subclasses may override this function to calculate something else.

**Review by TheaperDeng — COMMENTED** [review]
I am OK with these changes and will make another PR to implement them with solid test.

**Comment by jiaqima** [comment]
@TheaperDeng, one thing I'm not sure is if we want to set the default transform functions for both train and test as identical transformation? As in many cases, only test rep needs to be transformed.

**Comment by TheaperDeng** [comment]
> @TheaperDeng, one thing I'm not sure is if we want to set the default transform functions for both train and test as identical transformation? As in many cases, only test rep needs to be transformed.

It happens that I am working on this. I think one of the possible way is that, we just make `transform_test_rep` to be `abstractmethod`, that is users always need to implement this one. and for `transform_train_rep`, we can have a default implementation that returns the train rep directly.

**Comment by jiaqima** [comment]
> > @TheaperDeng, one thing I'm not sure is if we want to set the default transform functions for both train and test as identical transformation? As in many cases, only test rep needs to be transformed.
> 
> It happens that I am working on this. I think one of the possible way is that, we just make `transform_test_rep` to be `abstractmethod`, that is users always need to implement this one. and for `transform_train_rep`, we can have a default implementation that returns the train rep directly.
> 
> Or, we directly set them both to be defaultly return the rep directly. Users can override them when design their TDA methods.

I think the latter is Ok. Users implementing new attributors should be a bit more tolerant than users who just use the attributors, and they don't need an abstract method to remind them this is something they should override.

**Comment by TheaperDeng** [comment]
> > > @TheaperDeng, one thing I'm not sure is if we want to set the default transform functions for both train and test as identical transformation? As in many cases, only test rep needs to be transformed.
> > 
> > 
> > It happens that I am working on this. I think one of the possible way is that, we just make `transform_test_rep` to be `abstractmethod`, that is users always need to implement this one. and for `transform_train_rep`, we can have a default implementation that returns the train rep directly.
> > Or, we directly set them both to be defaultly return the rep directly. Users can override them when design their TDA methods.
> 
> I think the latter is Ok. Users implementing new attributors should be a bit more tolerant than users who just use the attributors, and they don't need an abstract method to remind them this is something they should override.

Yeah, I think so. And this makes the innerproduct IF can be directly instantiated and works like Grad-Dot, just like its name.

**Comment by jiaqima** [comment]
@tingwl0122 could you please take a look at the new implementation of the Arnoldi attributor and see if there is anything wrong?

**Comment by tingwl0122** [comment]
> @tingwl0122 could you please take a look at the new implementation of the Arnoldi attributor and see if there is anything wrong


I think the overall structure looks good to me. (Moving projection matrix computation to cache has been listed as one todo) Another thing is that the original Arnoldi paper actually used only a random batch of the full dataloader to do the algorithm, which I didn't include before. We can do something on the `data_target_pair` before feeding it into `func` to achieve this now. (also adding an additional param to control this random batch's size)

**Comment by tingwl0122** [comment]
> > @tingwl0122 could you please take a look at the new implementation of the Arnoldi attributor and see if there is anything wrong
> 
> I think the overall structure looks good to me. (Moving projection matrix computation to cache has been listed as one todo) Another thing is that the original Arnoldi paper actually used only a random batch of the full dataloader to do the algorithm, which I didn't include before. We can do something on the `data_target_pair` before feeding it into `func` to achieve this now. (also adding an additional param to control this random batch's size)

another minor thing is that we will need to force users to use `.cache` before attributing. We can probably raise some errors for this.

**Comment by jiaqima** [comment]
> > > @tingwl0122 could you please take a look at the new implementation of the Arnoldi attributor and see if there is anything wrong
> > 
> > 
> > I think the overall structure looks good to me. (Moving projection matrix computation to cache has been listed as one todo) Another thing is that the original Arnoldi paper actually used only a random batch of the full dataloader to do the algorithm, which I didn't include before. We can do something on the `data_target_pair` before feeding it into `func` to achieve this now. (also adding an additional param to control this random batch's size)
> 
> another minor thing is that we will need to force users to use `.cache` before attributing. We can probably raise some errors for this.

Sounds great. Thanks for the notes. I'll add them later.

**Comment by TheaperDeng** [comment]
> > > > @tingwl0122 could you please take a look at the new implementation of the Arnoldi attributor and see if there is anything wrong
> > > 
> > > 
> > > I think the overall structure looks good to me. (Moving projection matrix computation to cache has been listed as one todo) Another thing is that the original Arnoldi paper actually used only a random batch of the full dataloader to do the algorithm, which I didn't include before. We can do something on the `data_target_pair` before feeding it into `func` to achieve this now. (also adding an additional param to control this random batch's size)
> > 
> > 
> > another minor thing is that we will need to force users to use `.cache` before attributing. We can probably raise some errors for this.
> 
> Sounds great. Thanks for the notes. I'll add them later.

Thanks @tingwl0122 @jiaqima I think I can directly make the changes in my implementation #133

**Comment by jiaqima** [comment]
Closing this PR in favor of #133. Any follow-up discussion could be done there.
